File: android-agent/manual-build/src/com/hrapp/agent/MiniWebSocket.java
package com.hrapp.agent;

import java.io.*;
import java.net.Socket;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.util.Base64;

/**
 * ponytail: hand-rolled RFC 6455 client (unmasked-server / masked-client text
 * frames only, no fragmentation, no TLS) — written specifically to avoid
 * needing OkHttp+Kotlin-stdlib jars for this sandboxed manual-build proof.
 * The real Gradle project (../app) uses OkHttp; this class exists only in
 * manual-build/ and should NOT be treated as the long-term implementation —
 * swap back to OkHttp once `build.bat` can run (see docs/PROJECT_PROGRESS.md).
 * Ceiling: single-threaded blocking reads, one text frame per read() call.
 */
public class MiniWebSocket {
    public interface Listener {
        void onOpen();
        void onMessage(String text);
        void onFailure(Exception e);
        void onClosed();
    }

    private final String host;
    private final int port;
    private final String path;
    private final Listener listener;
    private Socket socket;
    private OutputStream out;
    private InputStream in;
    private volatile boolean running = true;

    public MiniWebSocket(String host, int port, String path, Listener listener) {
        this.host = host;
        this.port = port;
        this.path = path;
        this.listener = listener;
    }

    public void connectAsync() {
        new Thread(new Runnable() {
            @Override public void run() { runLoop(); }
        }, "mini-ws").start();
    }

    private void runLoop() {
        try {
            socket = new Socket(host, port);
            out = socket.getOutputStream();
            in = socket.getInputStream();
            handshake();
            listener.onOpen();
            readLoop();
        } catch (Exception e) {
            listener.onFailure(e);
        } finally {
            try { if (socket != null) socket.close(); } catch (IOException ignored) {}
            listener.onClosed();
        }
    }

    private void handshake() throws IOException {
        byte[] keyBytes = new byte[16];
        new SecureRandom().nextBytes(keyBytes);
        String key = Base64.getEncoder().encodeToString(keyBytes);

        String req = "GET " + path + " HTTP/1.1\r\n" +
                "Host: " + host + ":" + port + "\r\n" +
                "Upgrade: websocket\r\n" +
                "Connection: Upgrade\r\n" +
                "Sec-WebSocket-Key: " + key + "\r\n" +
                "Sec-WebSocket-Version: 13\r\n\r\n";
        out.write(req.getBytes(StandardCharsets.UTF_8));
        out.flush();

        BufferedReader reader = new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8));
        String line;
        boolean sawSwitchingProtocols = false;
        while ((line = reader.readLine()) != null && !line.isEmpty()) {
            if (line.startsWith("HTTP/1.1 101")) sawSwitchingProtocols = true;
        }
        if (!sawSwitchingProtocols) throw new IOException("WebSocket handshake failed");
        // Note: not validating Sec-WebSocket-Accept — acceptable for this
        // trusted-dev-relay demo, NOT for a production client.
    }

    public synchronized void send(String text) {
        try {
            byte[] payload = text.getBytes(StandardCharsets.UTF_8);
            byte[] mask = new byte[4];
            new SecureRandom().nextBytes(mask);
            ByteArrayOutputStream frame = new ByteArrayOutputStream();
            frame.write(0x81); // FIN + text frame opcode
            int len = payload.length;
            if (len <= 125) {
                frame.write(0x80 | len);
            } else if (len <= 65535) {
                frame.write(0x80 | 126);
                frame.write((len >> 8) & 0xFF);
                frame.write(len & 0xFF);
            } else {
                throw new IOException("payload too large for this demo client");
            }
            frame.write(mask);
            for (int i = 0; i < len; i++) {
                frame.write(payload[i] ^ mask[i % 4]);
            }
            out.write(frame.toByteArray());
            out.flush();
        } catch (IOException e) {
            listener.onFailure(e);
        }
    }

    private void readLoop() throws IOException {
        while (running) {
            int b0 = in.read();
            if (b0 == -1) return;
            int opcode = b0 & 0x0F;
            int b1 = in.read();
            boolean masked = (b1 & 0x80) != 0; // server frames are unmasked per spec, but handle both
            int len = b1 & 0x7F;
            if (len == 126) {
                len = (in.read() << 8) | in.read();
            } else if (len == 127) {
                for (int i = 0; i < 8; i++) in.read(); // skip extended len (demo scope: no >64KB frames)
                len = (in.read() << 8) | in.read();
            }
            byte[] maskKey = null;
            if (masked) {
                maskKey = new byte[4];
                for (int i = 0; i < 4; i++) maskKey[i] = (byte) in.read();
            }
            byte[] payload = new byte[len];
            int readTotal = 0;
            while (readTotal < len) {
                int r = in.read(payload, readTotal, len - readTotal);
                if (r == -1) return;
                readTotal += r;
            }
            if (masked) {
                for (int i = 0; i < len; i++) payload[i] ^= maskKey[i % 4];
            }
            if (opcode == 0x8) { running = false; return; } // close frame
            if (opcode == 0x1) { // text frame
                listener.onMessage(new String(payload, StandardCharsets.UTF_8));
            }
        }
    }

    public void close() {
        running = false;
        try { if (socket != null) socket.close(); } catch (IOException ignored) {}
    }
}

File: android-agent/app/src/main/java/com/hrapp/agent/MiniWebSocket.kt
package com.hrapp.agent

import java.io.BufferedReader
import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.io.InputStreamReader
import java.io.OutputStream
import java.net.Socket
import java.nio.charset.StandardCharsets
import java.security.SecureRandom
import java.util.Base64

/**
 * Minimal RFC 6455 WebSocket client (masked client text frames only, no TLS, no
 * fragmentation). Replaces OkHttp so the agent has zero third-party dependencies
 * and can be built without Gradle/AndroidX (ponytail — one fewer jar to resolve).
 * Ceiling: single blocking reader thread, one text frame per read; fine for the
 * relay's small JSON + base64 frames. Swap to OkHttp + wss:// for TLS in
 * production (MASTER.md §17).
 */
class MiniWebSocket(
    private val host: String,
    private val port: Int,
    private val path: String,
    private val listener: Listener
) {
    interface Listener {
        fun onOpen()
        fun onMessage(text: String)
        fun onFailure(t: Throwable)
        fun onClosed()
    }

    private var socket: Socket? = null
    private var out: OutputStream? = null
    private var input: InputStream? = null
    @Volatile private var running = true

    fun connectAsync() { Thread({ runLoop() }, "mini-ws").start() }

    private fun runLoop() {
        try {
            val s = Socket(host, port)
            socket = s
            out = s.getOutputStream()
            input = s.getInputStream()
            handshake()
            listener.onOpen()
            readLoop()
        } catch (e: Exception) {
            listener.onFailure(e)
        } finally {
            try { socket?.close() } catch (_: Exception) {}
            listener.onClosed()
        }
    }

    private fun handshake() {
        val keyBytes = ByteArray(16).also { SecureRandom().nextBytes(it) }
        val key = Base64.getEncoder().encodeToString(keyBytes)
        val req = "GET $path HTTP/1.1\r\n" +
            "Host: $host:$port\r\n" +
            "Upgrade: websocket\r\n" +
            "Connection: Upgrade\r\n" +
            "Sec-WebSocket-Key: $key\r\n" +
            "Sec-WebSocket-Version: 13\r\n\r\n"
        out!!.write(req.toByteArray(StandardCharsets.UTF_8)); out!!.flush()
        val reader = BufferedReader(InputStreamReader(input!!, StandardCharsets.UTF_8))
        var ok = false
        while (true) {
            val line = reader.readLine() ?: break
            if (line.isEmpty()) break
            if (line.startsWith("HTTP/1.1 101")) ok = true
        }
        if (!ok) throw Exception("WebSocket handshake failed")
        // Note: Sec-WebSocket-Accept not validated — trusted dev relay only.
    }

    @Synchronized
    fun send(text: String) {
        try {
            val o = out ?: return
            val payload = text.toByteArray(StandardCharsets.UTF_8)
            val mask = ByteArray(4).also { SecureRandom().nextBytes(it) }
            val frame = ByteArrayOutputStream()
            frame.write(0x81) // FIN + text
            val len = payload.size
            when {
                len <= 125 -> frame.write(0x80 or len)
                len <= 65535 -> { frame.write(0x80 or 126); frame.write((len shr 8) and 0xFF); frame.write(len and 0xFF) }
                else -> { frame.write(0x80 or 127); for (i in 7 downTo 0) frame.write((len ushr (8 * i)) and 0xFF) }
            }
            frame.write(mask)
            for (i in 0 until len) frame.write(payload[i].toInt() xor mask[i % 4].toInt())
            o.write(frame.toByteArray()); o.flush()
        } catch (e: Exception) { listener.onFailure(e) }
    }

    private fun readLoop() {
        val ins = input!!
        while (running) {
            val b0 = ins.read(); if (b0 == -1) return
            val opcode = b0 and 0x0F
            val b1 = ins.read(); if (b1 == -1) return
            val masked = (b1 and 0x80) != 0
            var len = b1 and 0x7F
            if (len == 126) { len = (ins.read() shl 8) or ins.read() }
            else if (len == 127) { repeat(6) { ins.read() }; len = (ins.read() shl 8) or ins.read() }
            val maskKey = if (masked) ByteArray(4) { ins.read().toByte() } else null
            val payload = ByteArray(len)
            var read = 0
            while (read < len) { val r = ins.read(payload, read, len - read); if (r == -1) return; read += r }
            if (masked && maskKey != null) for (i in 0 until len) payload[i] = (payload[i].toInt() xor maskKey[i % 4].toInt()).toByte()
            when (opcode) {
                0x8 -> { running = false; return } // close
                0x1 -> listener.onMessage(String(payload, StandardCharsets.UTF_8)) // text
            }
        }
    }

    fun close() {
        running = false
        try { socket?.close() } catch (_: Exception) {}
    }
}
